### ISSUE TITLE
What about providing some detailed examples to display how to run real jobs on the remote clusters?

### ISSUE THREAD

**Issue by RedStar-Iron** [Quantum-Accelerators/quacc#1018]
### How can we improve the documentation?

Hi Andrew,

Recently, I installed quacc and covalent on my own wsl2(Ubuntu), and I have used quacc to run some jobs locally. And when I tried to use quacc to run the VASP job on the "XSEDE-expanse", it failed, although I have installed covalent-hpc-plugin on my wsl and created a new conda enviroment on expanse where I installed covalent and psij-python. If there is a detailed example to introduce how to submit a vasp job on the remote cluster from the local terminal might be helpful.

**Comment by Andrew-S-Rosen** [comment]
Hi @RedStar-Iron! Thanks for your comment! I appreciate your report.

I agree that deploying remotely is the hardest part by far. I'm sure you have seen [this page](https://quantum-accelerators.github.io/quacc/user/wflow_engine/executors.html) already, which defined a sample HPC executor instance.

I imagine if you were following that, you actually were setting things up correctly. May I ask which version of quacc and Covalent are you using? There are some gnarly bugs with Covalent < 0.227.0rc0. The main branch of quacc has just been made compatible with Covalent >= 0.227.0rc0 (I'll mint a new release tonight).

But regardless, I agree that a complete example be fruitful. I'll add that shortly and make sure it works with the most recent versions of everything.

**Comment by RedStar-Iron** [comment]
Hi @Andrew-S-Rosen! Thanks for your reply!

By far, I used the quacc=0.3.1 and Covalent=0.226. When I tried to do a relax_job with VASP or even emt, I got an error  "Invalid OpenSSH certificate", but I already generate the id_rsa&id_rsa.pub and can ssh to the expanse cluster directly without typing password. And if I just did some easy calculation in the workflow, such as "2+2", it went smoothly.

**Comment by Andrew-S-Rosen** [comment]
@RedStar-Iron: Thanks for the reply. That sounds like it could be related to some of the prior Covalent bugs. I would try updating quacc (which will also update Covalent) and see if that fixes it. If not, it would be particularly helpful if you could share a full example (e.g. the EMT one) so I can troubleshoot. But my gut instinct is that you are following the instructions correctly and it's a bug related to setting inheritance in Covalent.

**Comment by RedStar-Iron** [comment]
Hi @Andrew-S-Rosen,

The code below is what I used to run emt job on expanse. And I have a question about how quacc and covalent started the vasp job on the remote cluster. I didn't supply the absolute path of vasp on the expanse to the quacc and covalent directly, did it use the ase env variable on the expanse `ASE_VASP_COMMAND & VASP_PP_PATH` ? 

And I set the `ASE_VASP_COMMAND="mpirun -np 32 path/to/vasp"` on expanse, while in this config file, `vasp_parallel_cmd = (f"srun -N {n_nodes} --ntasks-per-node={n_cores_per_node} --cpu_bind=cores")`, did it just extract the path of vasp from `ASE_VASP_COMMAND`? 

Also, I looked the code of `vasp.py`, it also called the Vasp class in ASE, so, did quacc call the ASE on expanse to run the vasp? I feel a bit confused about these questions. 

Maybe, if there is a flowchart to depict the principle of this workflow might be helpful to understand. 

```
import covalent as ct
from ase.build import bulk, molecule
from quacc import flow
from quacc import SETTINGS
from quacc.recipes.emt.core import relax_job


SETTINGS.VASP_USE_CUSTODIAN = False
n_nodes = 1  # Number of nodes to reserve for each calculation
n_cores_per_node = 32  # Number of CPU cores per node
vasp_parallel_cmd = (
    f"srun -N {n_nodes} --ntasks-per-node={n_cores_per_node} --cpu_bind=cores"
)

executor = ct.executor.HPCExecutor(
    # SSH credentials
    username="gsliu2021",
    address="login01.expanse.sdsc.edu",
    ssh_key_file="~/.ssh/id_rsa",
    cert_file="~/.ssh/id_rsa.pub",  
    # PSI/J parameters
    instance="slurm",
    resource_spec_kwargs={
        "node_count": n_nodes,
        "processes_per_node": n_cores_per_node,
    },  

    job_attributes_kwargs={
        "duration": 10,  # minutes
        "project_name": "uop149",
        "queue_name": "debug",
        # "custom_attributes": {"slurm.constraint": "cpu", "slurm.qos": "debug"},
    },  
    environment={"QUACC_VASP_PARALLEL_CMD": vasp_parallel_cmd},
    # Pre-/post-launch commands
    pre_launch_cmds=["module load purge","module load slurm"],
    # Remote Python env parameters
    remote_conda_env="finetuna",

    # Covalent parameters
    remote_workdir="/expanse/projects/qstore/uop149/gsliu2021/quacc",
    create_unique_workdir=True, 
    log_stdout="/expanse/projects/qstore/uop149/gsliu2021/quacc",
    log_stderr="/expanse/projects/qstore/uop149/gsliu2021/quacc",
)

@flow(executor=executor)
def workflow(atoms1, atoms2):
    # Define two independent relaxation jobs
    result1 = relax_job(atoms1)
    result2 = relax_job(atoms2)

    return {"result1": result1, "result2": result2}

# Define two Atoms objects
atoms1 = bulk("Cu")
atoms2 = molecule("N2")

# Dispatch the workflow to the Covalent server
dispatch_id = ct.dispatch(workflow)(atoms1, atoms2)

# Fetch the results from the server
result = ct.get_result(dispatch_id, wait=True)
print(result)
```

**Comment by Andrew-S-Rosen** [comment]
> The code below is what I used to run emt job on expanse. 

Thanks for the detailed report. I'll take a look on Friday (currently wrapping up vacation but had a few minutes to check GitHub!).

> and I have a question about how quacc and covalent started the vasp job on the remote cluster. I didn't supply the absolute path of vasp on the expanse to the quacc and covalent directly, did it use the ase env variable on the expanse ASE_VASP_COMMAND & VASP_PP_PATH

This is a good question. By default, quacc executes VASP by calling ASE but instead of having ASE run VASP directly, ASE runs Custodian for on-the-fly error handling. That is the section below:

https://github.com/Quantum-Accelerators/quacc/blob/d30c8f7e754831cad2632a32ee896aeb423c05f5/src/quacc/calculators/vasp.py#L251-L254

However, in your scenario, it looks like you are trying to disable the use of Custodian, in which case quacc will execute VASP via ASE as you are used to.

In the former case, that's where the `VASP_PARALLEL_CMD` and related settings come into play:

https://github.com/Quantum-Accelerators/quacc/blob/d30c8f7e754831cad2632a32ee896aeb423c05f5/src/quacc/calculators/custodian/vasp.py#L80-L87

In the latter case, quacc is calling ASE as usual and therefore `ASE_VASP_COMMAND` mut be set:

https://github.com/Quantum-Accelerators/quacc/blob/d30c8f7e754831cad2632a32ee896aeb423c05f5/src/quacc/calculators/vasp.py#L256-L260

This could likely be made clear in the [VASP setup](https://quantum-accelerators.github.io/quacc/install/codes.html#vasp) part of the instructions. Right now, it is written under the assumption that the user is going to run with Custodian, which is the default.

Anyway, to answer your question directly now: in your scenario, yes, it likely used your ASE environment variables (but ignored `QUACC_VASP_PARALLEL_CMD` altogether).

**Comment by RedStar-Iron** [comment]
Thanks for your detailed reply! I am much clearer about the process of quacc now, I will submit a slurm job and look into the code as well which can help me to further understand how the workflow works. Thanks again for your help!

**Comment by Andrew-S-Rosen** [comment]
@RedStar-Iron: Great! Keep me posted.

And as a note to myself, I am proposing the following additions to the docs:

- A section in the docs doing a top-to-bottom worked example for deploying calculations for several types of recipes, namely: EMT, ORCA, and VASP as prototypical examples. This should, ideally, be mostly copy-and-pasteable. It should also highlight how to approach this process gradually, e.g. first submitting a Python script as a slurm job, then using the workflow engine.

**Comment by Andrew-S-Rosen** [comment]
@RedStar-Iron: I updated the documentation per your request (see #1025 and the updated "Using a Workflow Engine" section of the live documentation).

I was able to confirm that the EMT workflow shared in the docs works as expected. However, I have confirmed that there is a bug in PSI/J, which is a dependency that `covalent-hpc-plugin` (i.e. `HPCExecutor`) relies on, and this causes an issue for VASP. See https://github.com/ExaWorks/psij-python/issues/423.
